### PR TITLE
ARROW-14753: [Doc] Steps in making your first PR - building C++

### PR DESCRIPTION
**Pull request by AlenkaF**
Add Building C++ and PyArrow sections of the New Contributor's guide.

### CODE REVIEW

File: docs/source/developers/guide/step_by_step/building.rst [hunk 1]
```diff
@@ -35,17 +35,83 @@
...
+to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
+`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
 
 Building C++
 ============
 
+Why build C++ from source?
+--------------------------
+
+The core of Arrow is written in C++ and all work done in other
```
Comment (rok) [inline docs/source/developers/guide/step_by_step/building.rst:54]:
```suggestion
The core of Arrow is written in C++ and all bindings in other
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 2]
```diff
@@ -35,17 +35,83 @@
...
+languages (Python, R, ..) is wrapping-binding-connecting underlying
```
Comment (rok) [inline docs/source/developers/guide/step_by_step/building.rst:55]:
```suggestion
languages (Python, R, ..) are wrapping underlying
```

**Review by rok — CHANGES_REQUESTED** [review]
Just some suggestions.

**Comment by AlenkaF** [comment]
Failing build doesn't seem to be related (RStudio Package Manager error).

File: docs/source/developers/guide/step_by_step/building.rst [hunk 3]
```diff
@@ -35,17 +35,83 @@
 Building Arrow's libraries 🏋🏿‍♀️
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:35]: It's not in this PR, but this should probably be "Building the Arrow libraries".

File: docs/source/developers/guide/step_by_step/building.rst [hunk 4]
```diff
@@ -35,17 +35,83 @@
...
 *********************************
 
+If you decide to contribute to Arrow you will meet the topic of
+compiling source code and use of CMake. You may have some
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:39]: The mention of CMake applies mostly to Arrow C++ and C++-derived bindings such as Python or R (not Java or Go, for example). Perhaps you could start this document by mentioning this context?

File: docs/source/developers/guide/step_by_step/building.rst [hunk 5]
```diff
@@ -35,17 +35,83 @@
...
+building Arrow better.
 
+If you feel comfortable with compiling then feel free to proceed
+to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
+`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
 
 Building C++
 ============
 
+Why build C++ from source?
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:51]: "Arrow C++", not just "C++" (and in the title above as well) :-)

File: docs/source/developers/guide/step_by_step/building.rst [hunk 6]
```diff
@@ -35,17 +35,83 @@
...
+The core of Arrow is written in C++ and all bindings in other
+languages (Python, R, ..) are wrapping underlying
+C++ functions. Even if you want to work on PyArrow or R package
+the source code of C++ may have to be edited also.
+
+About CMake
+-----------
+
+CMake is a cross platform build system generator and it uses make
+for the actual build. In the compiling process of Arrow what will
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:63]: Hmm, it doesn't necessarily use make (we actually recommend [Ninja](https://ninja-build.org/) in the C++ build docs).

Therefore you could rephrase this as : "... and it defers to another program such as `make` or `ninja` for the actual build".

File: docs/source/developers/guide/step_by_step/building.rst [hunk 7]
```diff
@@ -35,17 +35,83 @@
...
+most probably be needed is some tweaking of the flags that are added
+to cmake in the compiling process of Arrow.
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:65]: "CMake"

File: docs/source/developers/guide/step_by_step/building.rst [hunk 8]
```diff
@@ -35,17 +35,83 @@
...
+Detailed instructions on building C++ library from source can
+be found :ref:`here <building-arrow-cpp>`.
+
 .. _build-pyarrow:
 
 Building PyArrow
 ================
 
-.. _build-rarrow:
+After building C++ part of Arrow you have to build PyArrow on top of it
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:93]: "the Arrow C++ libraries"

File: docs/source/developers/guide/step_by_step/building.rst [hunk 9]
```diff
@@ -35,17 +35,83 @@
...
+also. The reason is the same, so you can edit the code and run tests on
+the edited code you have locally.
 
-Building R-Arrow
-================
+**Why do we have to do builds separately?**
```
Comment (pitrou) [inline docs/source/developers/guide/step_by_step/building.rst:97]: I'm not sure it's worth bothering beginners with this question :-) If we want to, though, we'll need to find a better wording for the answer, as currently it feels a bit confusing.

File: docs/source/developers/guide/step_by_step/building.rst (hunk 4 above)
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:39]:
```suggestion
compiling the source code and use of CMake. You may have some
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 10]
```diff
@@ -35,17 +35,83 @@
...
 
 Building C++
 ============
 
+Why build C++ from source?
+--------------------------
+
+The core of Arrow is written in C++ and all bindings in other
+languages (Python, R, ..) are wrapping underlying
+C++ functions. Even if you want to work on PyArrow or R package
```
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:56]: This is only true for the C++ implementation and its bindings. We should maybe try to reword this a bit to more correctly represent the fact that there are non-C++ Arrow implementations

File: docs/source/developers/guide/step_by_step/building.rst [hunk 11]
```diff
@@ -35,17 +35,83 @@
...
+the source code of C++ may have to be edited also.
+
+About CMake
+-----------
+
+CMake is a cross platform build system generator and it uses make
```
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:62]: The "uses make for the actual build" actually depends on the cmake settings. For example we recommend to use ninja instead of make.

File: docs/source/developers/guide/step_by_step/building.rst (hunk 6 above)
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:63]:
```suggestion
for the actual build. In the compilation process of Arrow what will
```

**Review by pitrou — COMMENTED** [review]
Some suggestions below. Do you intend to fill the TODOs here or in another PR?

File: docs/source/developers/guide/step_by_step/building.rst (hunk 7 above)
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:65]: This sentence is a bit hard to read. Maybe something like "You will probably need to tweak some CMake flags for compiling Arrow."

File: docs/source/developers/guide/step_by_step/building.rst (hunk 8 above)
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:93]:
```suggestion
After building the C++ part of Arrow you have to build PyArrow on top of it
```

**Review by jorisvandenbossche — COMMENTED** [review]
A very quick first pass!

File: docs/source/developers/guide/step_by_step/building.rst [hunk 12]
```diff
@@ -35,17 +35,83 @@
...
+for the actual build. In the compiling process of Arrow what will
+most probably be needed is some tweaking of the flags that are added
+to cmake in the compiling process of Arrow.
+
+
+Optional flags and why might we use them
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. TODO short description of the use of flags
```
Comment (wjones127) [inline docs/source/developers/guide/step_by_step/building.rst:70]: Two things worth mentioning here:

 1. R and Python have specific lists of flags in their respective building docs that can be referenced.
 2. We now have [CMake Presets](https://github.com/apache/arrow/blob/master/docs/source/developers/cpp/building.rst#cmake-presets) which are useful starting points.

File: docs/source/developers/guide/step_by_step/building.rst (hunk 4 above)
Comment (AlenkaF) [reply docs/source/developers/guide/step_by_step/building.rst:39]: Oh, didn't even know that. Will have to learn about other libraries to be able to explain the difference. Thanks for pointing it out!

Comment (AlenkaF) [reply docs/source/developers/guide/step_by_step/building.rst:39]: Well, it's nicely explained in https://arrow.apache.org/overview/ 🤦‍♀️

File: docs/source/developers/guide/step_by_step/building.rst (hunk 9 above)
Comment (AlenkaF) [reply docs/source/developers/guide/step_by_step/building.rst:97]: Yeah, showing I am still a little confused :P Will try to rephrase - as I would love to have this info when I started ;)

**Comment by AlenkaF** [comment]
Thanks everybody for the comments! 

> Do you intend to fill the TODOs here or in another PR?

I wanted to do it later, but I think I gained some understanding to manage it now. In case the review will take longer as expected I would still think about putting it into a separate PR to make the process faster.

As this is harder topic to write about for me (flags and env variables) it would be great to have some suggestions!

File: docs/source/developers/guide/step_by_step/building.rst [hunk 13]
```diff
@@ -31,21 +31,102 @@
...
+************************************
 
+The Arrow project contains good number of libraries that enable
+work in many languages. Most libraries (C++, C#, Go, Java,
+JavaScript, Julia, and Rust) already contain distinct implementations
+of Arrow. 
 
+It is different for C (Glib), MATLAB, Python, R, and Ruby as they
+are built on top of the C++ library. In this section we will be
+dealing with this second part of the libraries plus some of C++.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:45]:
```suggestion
dealing with some of these libraries as well has how they work with the C++ library.
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 14]
```diff
@@ -31,21 +31,102 @@
...
 
-Building C++
-============
+In this case, if you decide to contribute to Arrow you will meet
+the topic of compiling the source code and use of CMake. You may
+have some experience with it or not. If not, it is good to read
+through this part so you understand what is happening in the process
+of building Arrow better.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:51]:
```suggestion
If you decide to contribute to Arrow you might
need to compile the C++ source code.  This is done using a tool called CMake, which you may or may not have experience with.  If not, this section of the guide will help you better understand CMake and the process
of building Arrow's C++ code.
```

Not all contributors will need to compile the C++; in the R developer docs we have instructions for using the nightly builds as the C++ source when people are only making changes to the R code without needing to build C++ from source.

File: docs/source/developers/guide/step_by_step/building.rst [hunk 15]
```diff
@@ -31,21 +31,102 @@
...
+Building Arrow C++
+==================
+
+Why build Arrow C++ from source?
+--------------------------------
+
+For Arrow C++ implementation and its bindings (Python and R for example)
+these bindings are wrapping the underlying C++ functions. Even if you
+want to work on PyArrow or R package the source code of C++ may have to
+be edited also.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:66]:
```suggestion
For Arrow implementations which are built on top of the C++ implementation
(e.g. Python and R), wrappers and interfaces have been written to the 
underlying C++ functions. If you want to work on PyArrow or R package, 
you may need to edit the source code of the C++ library too.
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 16]
```diff
@@ -31,21 +31,102 @@
...
+be found :ref:`here <building-arrow-cpp>`.
 
 .. _build-pyarrow:
 
 Building PyArrow
 ================
 
-.. _build-rarrow:
+After building Arrow C++ part of Arrow you have to build PyArrow on top
+of it also. The reason is the same, so you can edit the code and run
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:109]:
```suggestion
of it also. The reason is the same; so you can edit the code and run
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 17]
```diff
@@ -31,21 +31,102 @@
...
+tests on the edited code you have locally.
 
-Building R-Arrow
-================
+**Why do we have to do builds separately?**
+
+As mentioned in the beginning of this page, Python part of the Arrow
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:114]:
```suggestion
As mentioned at the beginning of this page, the Python part of the Arrow
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 18]
```diff
@@ -31,21 +31,102 @@
...
+project is built on top of C++. In order to make changes in Python part
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:115]:
```suggestion
project is built on top of the C++ library. In order to make changes in the Python part
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 19]
```diff
@@ -31,21 +31,102 @@
...
+of Arrow as well as C++ part of Arrow, we need to build them separately.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:116]:
```suggestion
of Arrow as well as the C++ part of Arrow, you need to build them separately.
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 20]
```diff
@@ -31,21 +31,102 @@
...
+	Follow the instructions to build PyArrow together with the C++ library
+
+	- :ref:`build_pyarrow`
+	Or
+
+	- :ref:`build_pyarrow_win`
+
+.. _build-rapackage:
+
+Building R package
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:131]:
```suggestion
Building the R package
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 21]
```diff
@@ -31,21 +31,102 @@
...
+.. seealso::
+	Full list of optional flags: :ref:`cpp_build_optional_components`
+
+.. Environment variables useful for developers
+.. ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. TODO short description of the use of env vars
+
+Building from source vs. using binaries
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Using binaries is a fast and simple way of working with the latest
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply docs/source/developers/guide/step_by_step/building.rst:94]: Thanks for pointing it out! I will rephrase.

**Comment by AlenkaF** [comment]
Added the suggestions and the TODO parts (flags, env vars and CMake presets). I think this is ready for another round of review @thisisnic @jorisvandenbossche @pitrou Thank you!

File: docs/source/developers/guide/step_by_step/building.rst [hunk 22]
```diff
@@ -31,21 +31,120 @@
...
+paths with environment variables. In R the environment variables are used
+for all things connected to the build, also for setting CMake flags.
+
+Building from source vs. using binaries
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Using binaries is a fast and simple way of working with the last release
+of Arrow. However, if you use these it means that you will be unable to
+make changes to the Arrow C++ library.
+
+**Note:** every language has it's own way of dealing with binaries.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:119]:
```suggestion
**Note:** every language has its own way of dealing with binaries.
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 23]
```diff
@@ -31,21 +31,120 @@
...
+**********************************
 
+The Arrow project contains a number of libraries that enable
+work in many languages. Most libraries (C++, C#, Go, Java,
+JavaScript, Julia, and Rust) already contain distinct implementations
+of Arrow. 
 
+It is different for C (Glib), MATLAB, Python, R, and Ruby as they
+are built on top of the C++ library. In this section of the guide
+we will try to make a friendly introduction to the build
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:45]:
```suggestion
we will try to make a friendly introduction to the build,
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 24]
```diff
@@ -31,21 +31,120 @@
...
+Building Arrow C++
+==================
 
-.. _build-rarrow:
+Why build Arrow C++ from source?
+--------------------------------
+
+For Arrow implementations which are built on top of the C++ implementation
+(e.g. Python and R), wrappers and interfaces have been written to the
+underlying C++ functions. If you want to work on PyArrow or R package,
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:67]:
```suggestion
underlying C++ functions. If you want to work on PyArrow or the R package,
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 25]
```diff
@@ -31,21 +31,120 @@
...
+If you decide to contribute to Arrow you might need to compile the
+C++ source code. This is done using a tool called CMake, which you
+may or may not have experience with. If not, this section of the
+guide will help you better understand CMake and the process
+of building Arrow's C++ code.
 
-.. _build-pyarrow:
+If you feel comfortable with compiling then feel free to proceed
+to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
+`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
```
Comment (thisisnic) [inline docs/source/developers/guide/step_by_step/building.rst:57]:
```suggestion
This content is intended to help explain the concepts related to 
and tools required for building Arrow's C++ library from source.
If you are looking for the specific required steps, or already feel comfortable 
with compiling Arrow's C++ library, then feel free to proceed
to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
```
How about we add a little bit here to emphasise that this section is about *understanding* and isn't intended to provide concrete steps for the reader?

**Review by thisisnic — APPROVED** [review]
A few changes to suggest but otherwise looking good to me

File: docs/source/developers/guide/step_by_step/building.rst [hunk 26]
```diff
@@ -31,87 +31,120 @@
...
+**********************************
+Building the Arrow libraries 🏋🏿‍♀️
+**********************************
+
+The Arrow project contains a number of libraries that enable
+work in many languages. Most libraries (C++, C#, Go, Java,
+JavaScript, Julia, and Rust) already contain distinct implementations
+of Arrow. 
+
+It is different for C (Glib), MATLAB, Python, R, and Ruby as they
```
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:43]:
```suggestion
This is different for C (Glib), MATLAB, Python, R, and Ruby as they
```

File: docs/source/developers/guide/step_by_step/building.rst [hunk 27]
```diff
@@ -31,87 +31,120 @@
...
-to cmake in the compiling process of Arrow.
+CMake is a cross-platform build system generator and it defers
+to another program such as ``make`` or ``ninja`` for the actual build.
+If you are running into errors with the build process, the first thing to
+do is to look at the error message thoroughly and check the building
+documentation for any similar error advice. Also changing the CMake flags
+for compiling Arrow could be useful.
+
+CMake presets
+-------------
```
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:84]:
```suggestion
^^^^^^^^^^^^^^^^^^^
```

(then it's a subsection below "CMake"?)

File: docs/source/developers/guide/step_by_step/building.rst [hunk 28]
```diff
@@ -31,87 +31,120 @@
...
 	Full list of optional flags: :ref:`cpp_build_optional_components`
 
-.. Environment variables useful for developers
-.. ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-.. TODO short description of the use of env vars
+R and Python have specific lists of flags in their respective builds
+that need to be included. You can find the links at the end
+of this section.
+
+In general on Python side the options are set with CMake flags and
```
Comment (jorisvandenbossche) [inline docs/source/developers/guide/step_by_step/building.rst:109]:
```suggestion
In general on Python side, the options are set with CMake flags and
```
